File: bf.go
package main

import (
	"errors"
	"fmt"
	"io"
	"math"
	"os"
)

const name = "bf"

func main() {
	if err := run(os.Args[1:]); err != nil {
		fmt.Fprintf(os.Stderr, "%s: %s\n", name, err)
		os.Exit(1)
	}
}

func run(args []string) error {
	var (
		source []byte
		err    error
	)
	if len(args) == 0 {
		source, err = io.ReadAll(os.Stdin)
	} else {
		source, err = os.ReadFile(args[0])
	}
	if err != nil {
		return err
	}
	codes, err := parse(source)
	if err != nil {
		return err
	}
	return execute(codes)
}

const (
	incr        = 0 // + -
	next        = 1 // > <
	putc        = 2 // .
	getc        = 3 // ,
	jmpz        = 4 // [
	jmpnz       = 5 // ]
	setz        = 6 // [-]
	mult        = 7 // [->+>++<<]
	offset      = 3
	mask        = 1<<offset - 1
	offset_mult = offset + 8
)

func parse(source []byte) ([]int, error) {
	var codes []int
	var jmps []int
	for i := 0; i < len(source); i++ {
		switch source[i] {
		case '+':
			codes = add(codes, incr, 1)
		case '-':
			codes = add(codes, incr, -1)
		case '>':
			codes = add(codes, next, 1)
		case '<':
			codes = add(codes, next, -1)
		case '.':
			codes = append(codes, putc)
		case ',':
			codes = append(codes, getc)
		case '[':
			jmps = append(jmps, len(codes))
			codes = append(codes, jmpz)
		case ']':
			if len(jmps) == 0 {
				return nil, fmt.Errorf("unmatched ] at byte %d", i+1)
			}
			if j, l := jmps[len(jmps)-1], len(codes); j+2 == l &&
				codes[j+1] == incr|-1<<offset {
				codes = append(codes[:j], setz)
			} else if j+2 < l && (l-j)%2 == 1 && isMults(codes, j) {
				for k, p := j+2, 0; k < l-1; j, k = j+1, k+2 {
					p += codes[k] >> offset
					codes[j] = mult | p<<offset_mult | int(byte(codes[k+1]>>offset))<<offset
				}
				codes[j] = setz
				codes = codes[:j+1]
			} else {
				codes[j] |= l << offset
				codes = append(codes, jmpnz|j<<offset)
			}
			jmps = jmps[:len(jmps)-1]
		}
	}
	if len(jmps) != 0 {
		for i, depth := len(source)-1, 0; i >= 0; i-- {
			switch source[i] {
			case '[':
				if depth--; depth < 0 {
					return nil, fmt.Errorf("unmatched [ at byte %d", i+1)
				}
			case ']':
				depth++
			}
		}
	}
	return codes, nil
}

func add(codes []int, code int, diff int) []int {
	if len(codes) == 0 || codes[len(codes)-1]&mask != code {
		return append(codes, code|diff<<offset)
	}
	codes[len(codes)-1] += diff << offset
	return codes
}

func isMults(codes []int, start int) bool {
	var p int
	for i, k := 0, start+1; k < len(codes); i, k = i+1, k+2 {
		if codes[k]&mask == incr &&
			(i > 0 || codes[k]>>offset == -1) &&
			codes[k+1]&mask == next {
			p += codes[k+1] >> offset
		} else {
			return false
		}
	}
	return p == 0
}

func execute(codes []int) error {
	memory, pointer := []byte{0}, 0
	for i := 0; i < len(codes); i++ {
		switch c := codes[i]; c & mask {
		case incr:
			memory[pointer] += byte(c >> offset)
		case next:
			if pointer += c >> offset; pointer < 0 {
				return errors.New("negative address access")
			}
			for pointer >= len(memory) {
				memory = append(memory, 0)
			}
		case putc:
			if _, err := os.Stdout.Write(memory[pointer : pointer+1]); err != nil {
				return err
			}
		case getc:
			if _, err := os.Stdin.Read(memory[pointer : pointer+1]); err != nil {
				if err != io.EOF {
					return err
				}
				memory[pointer] = math.MaxUint8
			}
		case jmpz:
			if memory[pointer] == 0 {
				i = c >> offset
			}
		case jmpnz:
			if memory[pointer] != 0 {
				i = c >> offset
			}
		case setz:
			memory[pointer] = 0
		case mult:
			if v := memory[pointer]; v != 0 {
				dest := pointer + c>>offset_mult
				if dest < 0 {
					return errors.New("negative address access")
				}
				for dest >= len(memory) {
					memory = append(memory, 0)
				}
				memory[dest] += v * byte(c>>offset)
			}
		}
	}
	return nil
}

File: bf.rs
use std::io::{Read, Write};

const NAME: &str = "bf";

fn main() {
    if let Err(err) = run(std::env::args()) {
        eprintln!("{}: {}", NAME, err);
        std::process::exit(1)
    }
}

fn run(mut args: std::env::Args) -> Result<(), String> {
    let mut source = String::new();
    if let Some(file) = args.nth(1) {
        source = std::fs::read_to_string(file).map_err(|e| e.to_string())?;
    } else {
        std::io::stdin()
            .read_to_string(&mut source)
            .map_err(|e| e.to_string())?;
    }
    execute(parse(source)?)
}

const INCR: i64 = 0; // + -
const NEXT: i64 = 1; // > <
const PUTC: i64 = 2; // .
const GETC: i64 = 3; // ,
const JMPZ: i64 = 4; // [
const JMPNZ: i64 = 5; // ]
const SETZ: i64 = 6; // [-]
const MULT: i64 = 7; // [->+>++<<]
const OFFSET: i64 = 3;
const MASK: i64 = (1 << OFFSET) - 1;
const OFFSET_MULT: i64 = OFFSET + 8;

fn parse(source: String) -> Result<Vec<i64>, String> {
    let mut codes: Vec<i64> = Vec::new();
    let mut jmps: Vec<usize> = Vec::new();
    for (i, c) in source.chars().enumerate() {
        match c {
            '+' => add(&mut codes, INCR, 1),
            '-' => add(&mut codes, INCR, -1),
            '>' => add(&mut codes, NEXT, 1),
            '<' => add(&mut codes, NEXT, -1),
            '.' => codes.push(PUTC),
            ',' => codes.push(GETC),
            '[' => {
                jmps.push(codes.len());
                codes.push(JMPZ)
            }
            ']' => {
                if let Some(&j) = jmps.last() {
                    let l = codes.len();
                    if j + 2 == l && codes[j + 1] == INCR | (-1 << OFFSET) {
                        codes[j] = SETZ;
                        codes.resize(j + 1, 0);
                    } else if j + 2 < l
                        && (l - j) % 2 == 1
                        && codes[j + 1..]
                            .chunks(2)
                            .enumerate()
                            .try_fold(0, |p, (i, cs)| {
                                if cs[0] & MASK == INCR
                                    && (i > 0 || cs[0] >> OFFSET == -1)
                                    && cs[1] & MASK == NEXT
                                {
                                    Some(p + (cs[1] >> OFFSET))
                                } else {
                                    None
                                }
                            })
                            == Some(0)
                    {
                        let mut p: i64 = 0;
                        let mut j = j;
                        for k in (j + 2..l - 1).step_by(2) {
                            p += codes[k] >> OFFSET;
                            codes[j] = MULT
                                | p << OFFSET_MULT
                                | (((codes[k + 1] >> OFFSET) as u8) as i64) << OFFSET;
                            j += 1;
                        }
                        codes[j] = SETZ;
                        codes.resize(j + 1, 0);
                    } else {
                        codes[j] |= (l as i64) << OFFSET;
                        codes.push(JMPNZ | ((j as i64) << OFFSET));
                    }
                    jmps.pop();
                } else {
                    return Err(format!("unmatched ] at byte {}", i + 1));
                }
            }
            _ => {}
        }
    }
    if !jmps.is_empty() {
        let mut depth: i64 = 0;
        for (i, c) in source.chars().rev().enumerate() {
            match c {
                '[' => {
                    depth -= 1;
                    if depth < 0 {
                        return Err(format!("unmatched [ at byte {}", source.len() - i));
                    }
                }
                ']' => depth += 1,
                _ => {}
            }
        }
    }
    Ok(codes)
}

fn add(codes: &mut Vec<i64>, code: i64, diff: i64) {
    if codes.last().map_or(true, |c| c & MASK != code) {
        codes.push(code | (diff << OFFSET));
    } else if let Some(last) = codes.last_mut() {
        *last += diff << OFFSET;
    }
}

fn execute(codes: Vec<i64>) -> Result<(), String> {
    let mut memory: Vec<u8> = vec![0; 1];
    let mut pointer: usize = 0;
    let mut i = 0;
    while i < codes.len() {
        let c = codes[i];
        match c & MASK {
            INCR => {
                memory[pointer] = memory[pointer].wrapping_add((c >> OFFSET) as u8);
            }
            NEXT => {
                let dest = pointer as i64 + (c >> OFFSET);
                if dest < 0 {
                    return Err("negative address access".to_string());
                }
                pointer = dest as usize;
                if pointer >= memory.len() {
                    memory.resize(pointer + 1, 0);
                }
            }
            PUTC => {
                std::io::stdout()
                    .write(&memory[pointer..pointer + 1])
                    .map_err(|e| e.to_string())?;
            }
            GETC => {
                if let Err(e) = std::io::stdin().read_exact(&mut memory[pointer..pointer + 1]) {
                    if e.kind() == std::io::ErrorKind::UnexpectedEof {
                        return Ok(());
                    }
                    return Err(e.to_string());
                };
            }
            JMPZ => {
                if memory[pointer] == 0 {
                    i = (c >> OFFSET) as usize;
                }
            }
            JMPNZ => {
                if memory[pointer] != 0 {
                    i = (c >> OFFSET) as usize;
                }
            }
            SETZ => {
                memory[pointer] = 0;
            }
            MULT => {
                let v = memory[pointer];
                if v != 0 {
                    let dest = pointer as i64 + (c >> OFFSET_MULT);
                    if dest < 0 {
                        return Err("negative address access".to_string());
                    }
                    let dest = dest as usize;
                    if dest >= memory.len() {
                        memory.resize(dest + 1, 0);
                    }
                    memory[dest] = memory[dest].wrapping_add(v.wrapping_mul((c >> OFFSET) as u8));
                }
            }
            _ => {}
        }
        i += 1;
    }
    Ok(())
}
